HDDS-7543. ContainerStateMap uses HashSet to store container replica lists

cchung100m opened the pull request:
## What changes were proposed in this pull request?
ContainerStateMap uses HashSet to store container replica lists

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-7543

## How was this patch tested?
Existing unit and integration tests

sodonnel commented:
The original Jira states that the set can hold 3 replicas max, which is not true. In the case of EC 10-4, it will hold 14 replicas. If there is over replication it could even be a few more than that.

This change may be fine, but my initial thought is that we are going from a hash lookup to check for existence of a replica for additional or removal (container report processing, or dead node handling) to a list iteration, and therefore the runtime complexity is higher. Could this slow down container report processing?

Do we have any option to reduce the initial hashSet initial size to 6 or 8 to waste less memory?

I also haven't check where this code is called, but its unfortunate that a new hashSet needs to be constructed each time the replicas are retrieved. I understand why this is the case, as the list could otherwise be modified by container reporting or dead node handling after the copy has been retrieved. However, it would be relatively rare for the replicas for a container to change after the cluster has reached a stead state. Perhaps we can make the entries an un-modifiable set / list that can be returned to the caller directly, and then if the set / list needs to changed a new un-modifiable copy is created with the change to replace the original.

cchung100m commented:
Hi @sodonnel 

Thank you so much for the thoughtful and detailed feedback.

You're right regarding the replica count — the original Jira description was inaccurate in stating a max of 3 replicas. As you pointed out, in the case of EC 10-4, it can indeed go up to 14 or more with over-replication. I've updated the related code to reflect this more accurately.

I also agree with your concern about the performance impact when switching from a hash-based lookup to list iteration. To address this, I’ve revised the implementation to use a HashSet for replica existence checks, restoring constant-time lookup. This should help preserve efficiency, especially during container report processing and dead node handling.

Regarding the initial size of the HashSet, that's a great suggestion. I've adjusted the default initial capacity to 8, which seems like a good balance to reduce memory overhead without significantly impacting performance.

Thanks again for the great feedback — it’s helped make the solution cleaner and more robust. Let me know if you have any other thoughts!

> The original Jira states that the set can hold 3 replicas max, which is not true. In the case of EC 10-4, it will hold 14 replicas. If there is over replication it could even be a few more than that.
> 
> This change may be fine, but my initial thought is that we are going from a hash lookup to check for existence of a replica for additional or removal (container report processing, or dead node handling) to a list iteration, and therefore the runtime complexity is higher. Could this slow down container report processing?
> 
> Do we have any option to reduce the initial hashSet initial size to 6 or 8 to waste less memory?
> 
> I also haven't check where this code is called, but its unfortunate that a new hashSet needs to be constructed each time the replicas are retrieved. I understand why this is the case, as the list could otherwise be modified by container reporting or dead node handling after the copy has been retrieved. However, it would be relatively rare for the replicas for a container to change after the cluster has reached a stead state. Perhaps we can make the entries an un-modifiable set / list that can be returned to the caller directly, and then if the set / list needs to changed a new un-modifiable copy is created with the change to replace the original.

sodonnel commented:
I was thinking something like this would work:

```
--- a/hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/states/ContainerEntry.java
+++ b/hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/states/ContainerEntry.java
@@ -17,6 +17,7 @@
 
 package org.apache.hadoop.hdds.scm.container.states;
 
+import java.util.Collections;
 import java.util.HashMap;
 import java.util.HashSet;
 import java.util.Map;
@@ -31,7 +32,7 @@
  */
 public class ContainerEntry {
   private final ContainerInfo info;
-  private final Map<DatanodeID, ContainerReplica> replicas = new HashMap<>();
+  private Map<ContainerReplica, ContainerReplica> replicas = new HashMap<>();
 
   ContainerEntry(ContainerInfo info) {
     this.info = info;
@@ -42,14 +43,22 @@ public ContainerInfo getInfo() {
   }
 
   public Set<ContainerReplica> getReplicas() {
-    return new HashSet<>(replicas.values());
+    return Collections.unmodifiableSet(replicas.keySet());
   }
 
   public ContainerReplica put(ContainerReplica r) {
-    return replicas.put(r.getDatanodeDetails().getID(), r);
+    // Modifications are rare compared to reads, so to optimize for read we return and unmodifable view of the
+    // map in getReplicas. That means for any modification we need to copy the map.
+    Map<ContainerReplica, ContainerReplica> map = new HashMap<>(8);
+    map.putAll(this.replicas);
+    replicas = map;
+    return replicas.put(r, r);
   }
 
   public ContainerReplica removeReplica(DatanodeID datanodeID) {
+    Map<ContainerReplica, ContainerReplica> map = new HashMap<>(8);
+    map.putAll(this.replicas);
+    replicas = map;
     return replicas.remove(datanodeID);
   }
 }
```
However, what I am not sure about, is whether any of the caller of getReplicas() would assume they have their own copy they can modify, and therefore that code would fail. So while this would be efficient, it really constrains how the return value can be used, so it may be too risky to change.

cchung100m commented:
Hi @sodonnel 

Thanks for the prompt reply and suggestions. I can update the part you mentioned.

Hi @peterxcli 

Any thoughts would be appreciated if you are available. 😄 

>However, what I am not sure about, is whether any of the caller of getReplicas() would assume they have their own copy they can modify, and therefore that code would fail. So while this would be efficient, it really constrains how the return value can be used, so it may be too risky to change.

cchung100m commented:
Hi @jojochuang 

My username is cchung100m and I also leave a comment at jira ticket [HDDS-7544](https://issues.apache.org/jira/browse/HDDS-7544).

> @cchung100m do you have a jira account? I'd like to assigne the jira to you. thx

cchung100m commented:
Hi @peterxcli
Sodonnel gives me some suggestions but not sure whether the proposal is fit real case or not, do you have any thoughts on this? 😄

adoroszlai commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/states/ContainerEntry.java:
```diff
@@ -31,7 +31,8 @@
 public class ContainerEntry {
   private final ContainerInfo info;
-  private final Map<DatanodeID, ContainerReplica> replicas = new HashMap<>();
+  private Map<ContainerReplica, ContainerReplica> replicas = new HashMap<>();
```
`replicas` can be created lazily by the first modification, until then an empty map is fine.

```suggestion
  private Map<ContainerReplica, ContainerReplica> replicas = Collections.emptyMap();
```

adoroszlai commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/states/ContainerEntry.java:
```diff
@@ -42,14 +43,22 @@ public ContainerInfo getInfo() {
 
   public Set<ContainerReplica> getReplicas() {
-    return new HashSet<>(replicas.values());
+    return Collections.unmodifiableSet(replicas.keySet());
```
Since a new map is created for each change, we can create the unmodifiable wrapper on write instead.

adoroszlai commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/states/ContainerEntry.java:
```diff
@@ -31,8 +31,8 @@
-
+  private Map<ContainerReplica, ContainerReplica> replicaMap = Collections.emptyMap();
+  private Set<ContainerReplica> replicas = Collections.emptySet();
   ContainerEntry(ContainerInfo info) {
```
Please make sure to check [CI run](https://github.com/cchung100m/ozone/actions/runs/19934514493) in fork.  Checkstyle is [failing](https://github.com/cchung100m/ozone/actions/runs/19934514493/job/57155484251#step:16:17) with:

```
hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/states/ContainerEntry.java
 36: 'CTOR_DEF' should be separated from previous line.
```

adoroszlai reviewed (commented):
Thanks @cchung100m for iterating on this.

cchung100m commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/states/ContainerEntry.java:
```diff
@@ -42,14 +44,22 @@ public ContainerInfo getInfo() {
+    Map<DatanodeID, ContainerReplica> map = new HashMap<>(this.replicaMap);
+    T result = update.apply(map);
+    this.replicaMap = map;
+    this.replicas = Collections.unmodifiableSet(new HashSet<>(map.values()));
```
Hi @adoroszlai 

We might encounter race condition issue if we modified this.replicaMap directly:

> T result = update.apply(this.replicaMap); 
this.replicas = Collections.unmodifiableSet(new HashSet<>(this.replicaMap.values()));

adoroszlai replied:
`ContainerEntry` is used only by `ContainerStateManager` (via `ContainerStateMap`), while holding read or write lock.

https://github.com/apache/ozone/blob/0d4f6bbb2b01a441110c680d58f979d3284866f9/hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/ContainerStateManagerImpl.java#L407-L435

adoroszlai commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/states/ContainerEntry.java:
```diff
@@ -42,14 +44,20 @@ public ContainerInfo getInfo() {
+
+  private <T> T copyAndUpdate(java.util.function.Function<Map<DatanodeID, ContainerReplica>, T> update) {
+    T result = update.apply(this.replicaMap);
+    this.replicas = Collections.unmodifiableSet(new HashSet<>(this.replicaMap.values()));
```
nit: let's avoid `this.` like the rest of the class

```suggestion
    T result = update.apply(replicaMap);
    replicas = Collections.unmodifiableSet(new HashSet<>(replicaMap.values()));
```

cchung100m commented:
Thanks @adoroszlai
